crates/uv-resolver/src/lock.rs

```diff
@@ -716,6 +772,54 @@ impl SourceDist {
     }
 }
 
+/// Construct the lockfile-compatible [`URL`] for a [`GitSourceDist`].
+fn locked_git_url(git_dist: &GitSourceDist) -> Url {
+    let mut url = git_dist.git.repository().clone();
+
+    // Clear out any existing state.
+    url.set_fragment(None);
+    url.set_query(None);
+
+    // Put the subdirectory in the query.
+    if let Some(subdirectory) = git_dist.subdirectory.as_deref().and_then(Path::to_str) {
+        url.query_pairs_mut()
+            .append_pair("subdirectory", subdirectory);
+    }
+
+    // Put the requested reference in the query.
+    match git_dist.git.reference() {
+        GitReference::Branch(branch) => {
+            url.query_pairs_mut()
+                .append_pair("branch", branch.to_string().as_str());
+        }
+        GitReference::Tag(tag) => {
+            url.query_pairs_mut()
+                .append_pair("tag", tag.to_string().as_str());
+        }
+        GitReference::ShortCommit(rev)
+        | GitReference::BranchOrTag(rev)
+        | GitReference::BranchOrTagOrCommit(rev)
+        | GitReference::NamedRef(rev)
+        | GitReference::FullCommit(rev) => {
+            url.query_pairs_mut()
+                .append_pair("rev", rev.to_string().as_str());
+        }
+        GitReference::DefaultBranch => {}
+    }
```

Comment:
I feel like these need to be unified (part of what I was getting at in previous comments).

Reply:
Perhaps `GitReference` becomes internal to the crate, and that crate uses `GitSource` as its public interface? Not sure. It's useful (within the crate) to know that a ref can't be a commit, for example.

Reply:
Can you say more about what unification means in this context? I think as long as we require that a specific revision appears in the lock file (and the sub-directory if it exists), then I think that's all we actually need.

Reply:
Mostly just commenting that `GitReference` and `GitSource` are _nearly_ identical, except that `GitReference` contains more variants that break down `GitSource::Rev` in a way that's purely based on analyzing the value.